src/thunderbots/software/network_input/filter/robot_filter.cpp

```diff
@@ -1,22 +1,95 @@
 
-FilteredRobotData RobotFilter::getFilteredData(
+std::optional<Robot> RobotFilter::getFilteredData(
     const std::vector<SSLRobotDetection> &new_robot_data)
 {
+    int data_num               = 0;
+    Timestamp latest_timestamp = Timestamp().fromSeconds(0);
     FilteredRobotData filtered_data;
-    filtered_data.id               = 0;
-    filtered_data.position         = Point();
-    filtered_data.velocity         = Vector();
-    filtered_data.orientation      = Angle::zero();
-    filtered_data.angular_velocity = AngularVelocity::zero();
-    filtered_data.timestamp        = Timestamp::fromSeconds(0);
-
-    return filtered_data;
+
+    filtered_data.id               = this->getRobotId();
+    filtered_data.position         = Point(0, 0);
+    filtered_data.velocity         = Vector(0, 0);
+    filtered_data.orientation      = Angle().ofRadians(0);
+    filtered_data.angular_velocity = AngularVelocity().ofRadians(0);
+    filtered_data.timestamp        = Timestamp().fromSeconds(0);
+
+    for (const SSLRobotDetection &robot_data : new_robot_data)
+    {
+        // add up all data points for this robot and then average it
+        if (robot_data.id == this->getRobotId() &&
+            robot_data.timestamp > this->current_robot_state.lastUpdateTimestamp())
+        {
+            filtered_data.position = filtered_data.position + robot_data.position;
+            filtered_data.orientation =
+                filtered_data.orientation + robot_data.orientation;
+
+            filtered_data.timestamp = filtered_data.timestamp.fromMilliseconds(
+                filtered_data.timestamp.getMilliseconds() +
+                robot_data.timestamp.getMilliseconds());
+            data_num++;
+        }
+
+        // to get the latest timestamp of all data points in case there is no data for
+        // this robot id
+        if (latest_timestamp.getMilliseconds() < robot_data.timestamp.getMilliseconds())
+        {
+            latest_timestamp = robot_data.timestamp;
+        }
+    }
+
+    if (data_num == 0)
+    {
+        // if there is no data the duration of expiry_buffer_duration after previously
+        // recorded robot state, return null. Otherwise remain the same state
+        if (latest_timestamp.getMilliseconds() >
+            this->expiry_buffer_duration.getMilliseconds() +
+                current_robot_state.lastUpdateTimestamp().getMilliseconds())
+        {
+            return std::nullopt;
+        }
+        else
+        {
+            return std::make_optional(current_robot_state);
+        }
+    }
+    else
+    {
+        // update data by returning filtered robot data
+        filtered_data.position    = filtered_data.position / data_num;
+        filtered_data.orientation = filtered_data.orientation / data_num;
+
+        filtered_data.timestamp = filtered_data.timestamp.fromMilliseconds(
+            filtered_data.timestamp.getMilliseconds() / data_num);
+
+        // velocity = position difference/ time difference
```

Comment:
Nit:
Add a space after difference to match the comment below

Reply:
space added